packages/e2e-tests/test-utils/event-proxy-server.ts

```diff
@@ -0,0 +1,220 @@
+
+              listener(Buffer.from(JSON.stringify(data)).toString('base64'));
+            });
+            proxyResponse.end();
+          });
+
+          sentryResponse.addListener('error', err => {
+            throw err;
+          });
+
+          proxyResponse.writeHead(sentryResponse.statusCode || 500, sentryResponse.headers);
+        },
+      );
+
+      sentryRequest.write(Buffer.concat(proxyRequestChunks), 'binary');
+      sentryRequest.end();
+    });
+  });
+
+  const proxyServerStartupPromise = new Promise<void>(resolve => {
+    proxyServer.listen(options.port, () => {
+      resolve();
+    });
+  });
+
+  const eventCallbackServer = http.createServer((eventCallbackRequest, eventCallbackResponse) => {
+    eventCallbackResponse.statusCode = 200;
+    eventCallbackResponse.setHeader('connection', 'keep-alive');
+
+    const callbackListener = (data: string): void => {
+      eventCallbackResponse.write(data.concat('\n'), 'utf8');
+    };
+
+    eventCallbackListeners.add(callbackListener);
+
+    eventCallbackRequest.on('close', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+
+    eventCallbackRequest.on('error', () => {
+      eventCallbackListeners.delete(callbackListener);
+    });
+  });
+
+  const eventCallbackServerStartupPromise = new Promise<void>(resolve => {
+    const listener = eventCallbackServer.listen(0, () => {
+      const port = String((listener.address() as AddressInfo).port);
+      const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${options.proxyServerName}`);
+      fs.writeFileSync(tmpFileWithPort, port, { encoding: 'utf8' });
+      resolve();
+    });
+  });
+
+  await eventCallbackServerStartupPromise;
+  await proxyServerStartupPromise;
+  return;
+}
+
+export function waitForRequest(
+  proxyServerName: string,
+  callback: (eventData: SentryRequestCallbackData) => boolean,
+): Promise<SentryRequestCallbackData> {
+  const tmpFileWithPort = path.join(os.tmpdir(), `event-proxy-server-${proxyServerName}`);
+  const eventCallbackServerPort = fs.readFileSync(tmpFileWithPort, 'utf8');
+
+  return new Promise<SentryRequestCallbackData>((resolve, reject) => {
+    const request = http.request(`http://localhost:${eventCallbackServerPort}/`, {}, response => {
+      let eventContents = '';
+
+      response.on('error', err => {
+        reject(err);
+      });
+
+      response.on('data', (chunk: Buffer) => {
+        const chunkString = chunk.toString('utf8');
+        chunkString.split('').forEach(char => {
+          if (char === '\n') {
+            const eventCallbackData: SentryRequestCallbackData = JSON.parse(
+              Buffer.from(eventContents, 'base64').toString('utf8'),
+            );
+            if (callback(eventCallbackData)) {
+              response.destroy();
+              resolve(eventCallbackData);
+            }
+            eventContents = '';
+          } else {
+            eventContents = eventContents.concat(char);
```

Comment:
l: why `concat` over `push`? push is faster, and we already break mutability by using `let eventContents`

Reply:
Sadly string doesn't have a push method.

Reply:
omg im a dummy